Failures with AVX512 in numpy test suite when using blis on windows (through conda-forge)

Conda-forge is the community driven channel for the packaging & environment tool `conda`, which is used heavily in many scientific & corporate environments, especially (but not only) in terms of python.

conda-forge implements a blas [meta-package](https://github.com/conda-forge/blas-feedstock/blob/master/recipe/meta.yaml) that allows selecting between different blas/lapack implementations. Currently available are: netlib, mkl, openblas & blis (where blis uses the netlib binaries for lapack).

For the last couple of releases, I've been running the numpy (& scipy) test-suites against all those blas variants to root out any potential errors. Blis on windows (at least as packaged through conda-forge) was always a problem child, but has gotten a bit better recently. Still, I keep getting some ~flaky~ test failures both for the numpy & scipy test suites ~- flaky as in: they appear in one run, and then disappear again, with the only difference being a CI restart.~ due to absence/presence of AVX512.

Those errors look pretty scary numerically - e.g. all NaNs (where non-NaN would be expected), or a matrix that's completely zero instead of the identity - more details below.

For some example CI runs see [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344303&view=results) and [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=344255&view=results) (coming from https://github.com/conda-forge/numpy-feedstock/pull/227, but also occurring in https://github.com/conda-forge/numpy-feedstock/pull/237).

Reproducer (on a windows system with AVX512):
```
# install miniconda, e.g. from https://github.com/conda-forge/miniforge
conda config --add channels conda-forge
# set up environment
conda create -n test_env numpy "blas =*=blis" pytest hypothesis setuptools
# activate environment
conda activate test_env
# confirm that instruction sets '... AVX512F* AVX512CD* AVX512_SKX* ...' are detected
python -c "import numpy; numpy._pytesttester._show_numpy_info()"
# run (relevant subset of) the test suite
pytest --pyargs numpy.linalg.tests.test_linalg -v
```

<details>
<summary>Short log of failures</summary>

```
=========================== short test summary info ===========================
FAILED core/tests/test_multiarray.py::TestMatmul::test_dot_equivalent[args4]
FAILED linalg/tests/test_linalg.py::TestSolve::test_sq_cases - AssertionError...
FAILED linalg/tests/test_linalg.py::TestSolve::test_generalized_sq_cases - As...
FAILED linalg/tests/test_linalg.py::TestInv::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestInv::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_sq_cases - Ass...
FAILED linalg/tests/test_linalg.py::TestPinv::test_generalized_nonsq_cases - ...
FAILED linalg/tests/test_linalg.py::TestDet::test_sq_cases - AssertionError: ...
FAILED linalg/tests/test_linalg.py::TestDet::test_generalized_sq_cases - Asse...
FAILED linalg/tests/test_linalg.py::TestMatrixPower::test_power_is_minus_one[dt13]
FAILED linalg/tests/test_linalg.py::TestCholesky::test_basic_property - Asser...
= 11 failed, 13581 passed, 714 skipped, 1 deselected, 20 xfailed, 1 xpassed, 229 warnings in 447.85s (0:07:27) =
```

</details>


<details>
<summary>Failure of <code>TestMatmul.test_dot_equivalent[args4]</code></summary>

```
____________________ TestMatmul.test_dot_equivalent[args4] ____________________

self = <numpy.core.tests.test_multiarray.TestMatmul object at 0x000002A1BE3DFFD0>
args = (array([[ 0.,  1.,  2.],
       [ 3.,  4.,  5.],
       [ 6.,  7.,  8.],
       [ 9., 10., 11.],
       [12., 13., 14.]]), array([[ 0.,  3.,  6.,  9., 12.],
       [ 1.,  4.,  7., 10., 13.],
       [ 2.,  5.,  8., 11., 14.]]))

    @pytest.mark.parametrize('args', (
            [...]
        ))
    def test_dot_equivalent(self, args):
        r1 = np.matmul(*args)
        r2 = np.dot(*args)
>       assert_equal(r1, r2)
E       AssertionError:
E       Arrays are not equal
E
E       x and y nan location mismatch:
E        x: array([[nan, nan, nan, nan, nan],
E              [nan, nan, nan, nan, nan],
E              [nan, nan, nan, nan, nan],...
E        y: array([[  5.,  14.,  23.,  32.,  41.],
E              [ 14.,  50.,  86., 122., 158.],
E              [ 23.,  86., 149., 212., 275.],...
```

</details>

<details>
<summary>Failure of <code>TestSolve.test_sq_cases</code></summary>

```
___________________________ TestSolve.test_sq_cases ___________________________

actual = array([2.+1.j, 1.+2.j], dtype=complex64)
desired = array([0.+0.j, 0.+0.j], dtype=complex64), decimal = 6, err_msg = ''
verbose = True
```

</details>

<details>
<summary>Failure of <code>TestSolve.test_generalized_sq_cases</code></summary>

```
_____________________ TestSolve.test_generalized_sq_cases _____________________

actual = array([[ 2. +1.j,  1. +2.j],
       [14. +7.j,  7.+14.j],
       [12. +6.j,  6.+12.j]], dtype=complex64)
desired = array([[0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j]], dtype=complex64)
decimal = 6, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestInv.test_sq_cases</code></summary>

```
____________________________ TestInv.test_sq_cases ____________________________

actual = array([[0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j]], dtype=complex64)
desired = array([[1., 0.],
       [0., 1.]]), decimal = 6, err_msg = ''
verbose = True
```

</details>

<details>
<summary>Failure of <code>TestInv.test_generalized_sq_cases</code></summary>

```
______________________ TestInv.test_generalized_sq_cases ______________________

actual = array([[[0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j]],

       [[0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j]],

       [[0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j]]], dtype=complex64)
desired = array([[[1.+0.j, 0.+0.j],
        [0.+0.j, 1.+0.j]],

       [[1.+0.j, 0.+0.j],
        [0.+0.j, 1.+0.j]],

       [[1.+0.j, 0.+0.j],
        [0.+0.j, 1.+0.j]]], dtype=complex64)
decimal = 6, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestPinv.test_generalized_sq_cases</code></summary>

```
_____________________ TestPinv.test_generalized_sq_cases ______________________

[...]

self = <numpy.linalg.tests.test_linalg.TestPinv object at 0x000002A1C2382E50>
a = array([[[0.19151945, 0.62210877, 0.43772774, 0.78535858, 0.77997581,
         0.27259261, 0.27646426, 0.80187218],
   ...],
        [2.3715724 , 2.9762444 , 2.87640529, 2.37589241, 0.85575288,
         1.87475012, 1.43428139, 0.58702554]]])
b = array([[0.38231745, 0.05387369, 0.45164841, 0.98200474, 0.1239427 ,
        0.1193809 , 0.73852306, 0.58730363],
     ...543],
       [2.29390471, 0.32324211, 2.70989045, 5.89202845, 0.7436562 ,
        0.71628539, 4.43113834, 3.5238218 ]])
tags = frozenset({'generalized', 'square', 'strided'})

    def do(self, a, b, tags):
        a_ginv = linalg.pinv(a)
        # `a @ a_ginv == I` does not hold if a is singular
        dot = dot_generalized
>       assert_almost_equal(dot(dot(a, a_ginv), a), a, single_decimal=5, double_decimal=11)

[...]

actual = array([[[nan, nan, nan, nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan, nan, nan, nan],
        [nan, nan,..., nan, nan, nan],
        [nan, nan, nan, nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan, nan, nan, nan]]])
desired = array([[[0.19151945, 0.62210877, 0.43772774, 0.78535858, 0.77997581,
         0.27259261, 0.27646426, 0.80187218],
   ...],
        [2.3715724 , 2.9762444 , 2.87640529, 2.37589241, 0.85575288,
         1.87475012, 1.43428139, 0.58702554]]])
decimal = 11, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestPinv.test_generalized_nonsq_cases</code></summary>

```
____________________ TestPinv.test_generalized_nonsq_cases ____________________

[...]

self = <numpy.linalg.tests.test_linalg.TestPinv object at 0x000002A1C2395E80>
a = array([[[0.22921857, 0.89996519, 0.41675354, 0.53585166, 0.00620852,
         0.30064171, 0.43689317, 0.612149  , 0.91...8, 2.32854217, 2.34743398,
         2.28481174, 2.74320934, 1.97586835, 1.70510274, 0.60526708,
         2.09488913]]])
b = array([[0.95219541, 0.88996329, 0.99356736, 0.81870351, 0.54512217,
        0.45125405, 0.89055719, 0.97326479],
     ...354],
       [5.71317246, 5.33977972, 5.96140418, 4.91222106, 3.270733  ,
        2.70752433, 5.34334313, 5.83958875]])
tags = frozenset({'generalized', 'nonsquare', 'strided'})

    def do(self, a, b, tags):
        a_ginv = linalg.pinv(a)
        # `a @ a_ginv == I` does not hold if a is singular
        dot = dot_generalized
>       assert_almost_equal(dot(dot(a, a_ginv), a), a, single_decimal=5, double_decimal=11)

[...]

actual = array([[[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan, nan, nan, nan, nan,..., nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]]])
desired = array([[[0.22921857, 0.89996519, 0.41675354, 0.53585166, 0.00620852,
         0.30064171, 0.43689317, 0.612149  , 0.91...8, 2.32854217, 2.34743398,
         2.28481174, 2.74320934, 1.97586835, 1.70510274, 0.60526708,
         2.09488913]]])
decimal = 11, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestDet.test_sq_cases</code></summary>

```
____________________________ TestDet.test_sq_cases ____________________________

actual = (6-17j), desired = (3.9968028886505635e-15-4.000000000000001j)
decimal = 6, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestDet.test_generalized_sq_cases</code></summary>

```
______________________ TestDet.test_generalized_sq_cases ______________________

actual = array([ 6. -17.j, 24. -68.j, 54.-153.j], dtype=complex64)
desired = array([3.99680289e-15 -4.j, 1.59872116e-14-16.j, 2.13162821e-14-36.j])
decimal = 6, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestMatrixPower.test_power_is_minus_one[dt13]</code></summary>

```
________________ TestMatrixPower.test_power_is_minus_one[dt13] ________________

actual = array([[0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j]], dtype=complex64)
desired = array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])
decimal = 6, err_msg = '', verbose = True
```

</details>

<details>
<summary>Failure of <code>TestCholesky::test_basic_property</code></summary>

```
______________________ TestCholesky.test_basic_property _______________________

self = <numpy.linalg.tests.test_linalg.TestCholesky object at 0x000002A1C0E0A940>

    def test_basic_property(self):
        # Check A = L L^H
        shapes = [(1, 1), (2, 2), (3, 3), (50, 50), (3, 10, 10)]
        dtypes = (np.float32, np.float64, np.complex64, np.complex128)

        for shape, dtype in itertools.product(shapes, dtypes):
            np.random.seed(1)
            a = np.random.randn(*shape)
            if np.issubdtype(dtype, np.complexfloating):
                a = a + 1j*np.random.randn(*shape)

            t = list(range(len(shape)))
            t[-2:] = -1, -2

            a = np.matmul(a.transpose(t).conj(), a)
            a = np.asarray(a, dtype=dtype)

            c = np.linalg.cholesky(a)

            b = np.matmul(c, c.transpose(t).conj())
>           assert_allclose(b, a,
                            err_msg=f'{shape} {dtype}\n{a}\n{c}',
                            atol=500 * a.shape[0] * np.finfo(dtype).eps)
E           AssertionError:
E           Not equal to tolerance rtol=1e-07, atol=0.000119209
E           (2, 2) <class 'numpy.complex64'>
E           [[ 6.7107615-6.181477e-17j -3.746924 -9.348988e-01j]
E            [-3.746924 +9.348988e-01j  7.402024 -6.601837e-17j]]
E           [[2.5905137+0.j 0.       +0.j]
E            [0.       -0.j 2.7206662+0.j]]
E           Mismatched elements: 2 / 4 (50%)
E           Max absolute difference: 3.8617969
E           Max relative difference: 1.
E            x: array([[6.710761+0.j, 0.      +0.j],
E                  [0.      +0.j, 7.402024+0.j]], dtype=complex64)
E            y: array([[ 6.710762-6.181477e-17j, -3.746924-9.348988e-01j],
E                  [-3.746924+9.348988e-01j,  7.402024-6.601837e-17j]],
E                 dtype=complex64)
```

</details>

For example, just chasing down the first failure:
```
        r1 = np.matmul(*args)
        r2 = np.dot(*args)
>       assert_equal(r1, r2)
E       AssertionError:
E       Arrays are not equal
```
leads all over the place in the code base (with lots of templating etc. in between)
https://github.com/numpy/numpy/blob/v1.21.0/numpy/core/multiarray.py#L736
https://github.com/numpy/numpy/blob/v1.21.0/numpy/core/src/umath/matmul.c.src#L398
https://github.com/numpy/numpy/blob/v1.21.0/numpy/core/src/multiarray/arraytypes.c.src#L3536
etc.

and it's hard to work out which path gets taken exactly in order to construct a C-reproducer (ultimately for each of the failures...). Not sure if @seberg @charris @eric-wieser have some cycles to spare, but I doubt it.

> See if you can get this test program to run in the AVX512 Azure environment:

Doing this in https://github.com/conda-forge/blis-feedstock/pull/23, seems more/different headers are necessary?
```
+ /home/[...]/bin/x86_64-conda-linux-gnu-cc -I/home/[...]/include debug_blis.c
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_round':
debug_blis.c:(.text+0x2c9): undefined reference to `round'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_thread_broadcast':
debug_blis.c:(.text+0xe956): undefined reference to `bli_thrcomm_bcast'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_thread_barrier':
debug_blis.c:(.text+0xe97e): undefined reference to `bli_thrcomm_barrier'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_thread_range_jrir_sl':
debug_blis.c:(.text+0xea1d): undefined reference to `bli_thread_range_sub'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_membrk_init_mutex':
debug_blis.c:(.text+0x1071f): undefined reference to `bli_pthread_mutex_init'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_membrk_finalize_mutex':
debug_blis.c:(.text+0x10740): undefined reference to `bli_pthread_mutex_destroy'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_membrk_lock':
debug_blis.c:(.text+0x10821): undefined reference to `bli_pthread_mutex_lock'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_membrk_unlock':
debug_blis.c:(.text+0x10842): undefined reference to `bli_pthread_mutex_unlock'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_apool_lock':
debug_blis.c:(.text+0x10b97): undefined reference to `bli_pthread_mutex_lock'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_apool_unlock':
debug_blis.c:(.text+0x10bba): undefined reference to `bli_pthread_mutex_unlock'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `bli_gemmsup_ref_var1n2m_opt_cases':
debug_blis.c:(.text+0x1144d): undefined reference to `bli_abort'
/home/[...]/bin/ld: debug_blis.c:(.text+0x119c8): undefined reference to `bli_obj_imag_is_zero'
/home/[...]/bin/ld: /tmp/ccyxRQDz.o: in function `main':
debug_blis.c:(.text+0x11c6a): undefined reference to `bli_dgemm'
collect2: error: ld returned 1 exit status
```

No discernible output when calling the reproducer, but again there _is_ debug output from the selftests (I had removed the other debug statement for this run, so there is only one possible source). Note also that `check-blastest.sh` produces a failure that gets ignored somehow.

<details>
<summary>selftests</summary>

```
Running test_libblis.x (fast) with output redirected to 'output.testsuite'
Running cblat1.x > 'out.cblat1'
Running cblat2.x < './blastest/input/cblat2.in' (output to 'out.cblat2')
Running cblat3.x < './blastest/input/cblat3.in' (output to 'out.cblat3')
Running dblat1.x > 'out.dblat1'
Running dblat2.x < './blastest/input/dblat2.in' (output to 'out.dblat2')
Running dblat3.x < './blastest/input/dblat3.in' (output to 'out.dblat3')
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 1 0x1.0000000000000p+0 0x1.0000000000000p+0
1 1 0x1.0000000000000p+0 0x1.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 1 0x1.0000000000000p+0 0x1.0000000000000p+0
1 1 0x1.0000000000000p+0 0x1.0000000000000p+0
Running sblat1.x > 'out.sblat1'
Running sblat2.x < './blastest/input/sblat2.in' (output to 'out.sblat2')
Running sblat3.x < './blastest/input/sblat3.in' (output to 'out.sblat3')
Running zblat1.x > 'out.zblat1'
Running zblat2.x < './blastest/input/zblat2.in' (output to 'out.zblat2')
Running zblat3.x < './blastest/input/zblat3.in' (output to 'out.zblat3')
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
1 0 0x1.0000000000000p+0 0x0.0000000000000p+0
check-blastest.sh: At least one BLAS test failed. :(
check-blastest.sh: Please see out.* files for details.
make: [checkblas] Error 1 (ignored)
Makefile:778: recipe for target 'checkblas' failed
check-blistest.sh: All BLIS tests passed!
```

</details>

Regarding the discussion of the configuration above, not sure I understand all the differences between configuring for `x86_64` vs `skx`, but the latter seems to be active also for `x86_64`:

<details>
<summary>Excerpt of <code>configure</code></summary>

```
[...]
configure: starting configuration of BLIS 0.8.1.
configure: configuring with official version string.
configure: found shared library .so version '3.0.0'.
configure:   .so major version: 3
configure:   .so minor.build version: 0.0
configure: manual configuration requested; configuring with 'x86_64'.
configure: checking configuration against contents of 'config_registry'.
configure: configuration 'x86_64' is registered.
configure: 'x86_64' is defined as having the following sub-configurations:
configure:    skx knl haswell sandybridge penryn zen2 zen excavator steamroller piledriver bulldozer generic
configure: which collectively require the following kernels:
configure:    skx knl sandybridge penryn zen2 haswell zen piledriver bulldozer generic
configure: checking sub-configurations:
configure:   'skx' is registered...and exists.
configure:   'knl' is registered...and exists.
configure:   'haswell' is registered...and exists.
configure:   'sandybridge' is registered...and exists.
configure:   'penryn' is registered...and exists.
configure:   'zen2' is registered...and exists.
configure:   'zen' is registered...and exists.
configure:   'excavator' is registered...and exists.
configure:   'steamroller' is registered...and exists.
configure:   'piledriver' is registered...and exists.
configure:   'bulldozer' is registered...and exists.
configure:   'generic' is registered...and exists.
configure: checking sub-configurations' requisite kernels:
configure:   'skx' kernels...exist.
configure:   'knl' kernels...exist.
configure:   'sandybridge' kernels...exist.
configure:   'penryn' kernels...exist.
configure:   'zen2' kernels...exist.
configure:   'haswell' kernels...exist.
configure:   'zen' kernels...exist.
configure:   'piledriver' kernels...exist.
configure:   'bulldozer' kernels...exist.
configure:   'generic' kernels...exist.
configure: detected --prefix='/d/bld/blis_1625517330860/_h_env/Library'.
[...]
```

Hm, I had copied this from your previous print statement, but the signature (resp.) meaning of beta must be different here:
```
In file included from config/excavator/bli_cntx_init_excavator.c:35:
include/x86_64/blis.h: In function 'bli_dddxpbys_mxn':
include/x86_64/blis.h:15548:59: error: invalid operands to binary * (have 'char *' and 'double *')
15548 |  printf("bli_dddxpbys_mxn, before beta=0 branch; %g %a\n" *(double* restrict)beta, *(double* restrict)beta);
```

@fgvanzee it was buried up above. `make check` on Windows/AVX512 produces:

```
check-blastest.sh: At least one BLAS test failed. :(
check-blastest.sh: Please see out.* files for details.
make: [checkblas] Error 1 (ignored)
Makefile:778: recipe for target 'checkblas' failed
check-blistest.sh: All BLIS tests passed!
```

However, on a second look it doesn't actually seem that any failure is ignored. The make rule fails which I assume leads to a non-zero exit code. The "Error 1 (ignored)" is probably because `make -k` or similar is used?

BTW for context those failures registered as e.g. `******* FATAL ERROR - PARAMETER NUMBER 11 WAS CHANGED INCORRECTLY *******`. The parameter in question is always beta, so it seems that beta is getting corrupted somehow. This fits with the observed failure which was the original issue.

--- a/config/skx/make_defs.mk
+++ b/config/skx/make_defs.mk
@@ -71,7 +71,15 @@ ifeq ($(CC_VENDOR),icc)
 CKVECFLAGS     := -xCORE-AVX512
 else
 ifeq ($(CC_VENDOR),clang)
+# NOTE: We have to use -march=haswell on Windows because apparently AVX512
+# uses an alternate calling convention where xmm registers are not callee-saved
+# on the stack. When this is mixed with framework code compiled for general
+# x86_64 mode then chaos ensues (e.g. #514).
+ifeq ($(IS_WIN),yes)
+CKVECFLAGS     := -mavx512f -mavx512dq -mavx512bw -mavx512vl -mfpmath=sse -march=haswell
+else
 CKVECFLAGS     := -mavx512f -mavx512dq -mavx512bw -mavx512vl -mfpmath=sse -march=skylake-avx512
+endif
 else
 $(error gcc, icc, or clang is required for this configuration.)
 endif
@@ -97,7 +105,15 @@ ifeq ($(CC_VENDOR),icc)
 CRVECFLAGS     := -xCORE-AVX2
 else
 ifeq ($(CC_VENDOR),clang)
+# NOTE: We have to use -march=haswell on Windows because apparently AVX512
+# uses an alternate calling convention where xmm registers are not callee-saved
+# on the stack. When this is mixed with framework code compiled for general
+# x86_64 mode then chaos ensues (e.g. #514).
+ifeq ($(IS_WIN),yes)
+CRVECFLAGS     := -march=haswell -funsafe-math-optimizations -ffp-contract=fast
+else
 CRVECFLAGS     := -march=skylake-avx512 -mno-avx512f -mno-avx512vl -mno-avx512bw -mno-avx512dq -mno-avx512cd -funsafe-math-optimizations -ffp-contract=fast
+endif
 else
 $(error gcc, icc, or clang is required for this configuration.)
 endif

